Spatial_searching/include/CGAL/Fuzzy_sphere.h

```diff
@@ -32,134 +33,127 @@
-		}
-		
-		return (distance < squared_radius); 
-        }
-
-    bool inner_range_intersects(const Kd_tree_rectangle<FT,Dimension>& rectangle) const {
-      // test whether the interior of a sphere
-      // with radius (r-eps) intersects 'rectangle', i.e.
-      // if the minimal distance of c to 'rectangle' is less than r-eps
-      FT distance = FT(0);
-      FT squared_radius = (r-eps)*(r-eps);
-      typename SearchTraits::Construct_cartesian_const_iterator_d construct_it=
-                  traits.construct_cartesian_const_iterator_d_object();
-                typename SearchTraits::Cartesian_const_iterator_d cit = construct_it(c),
-                                                                  end = construct_it(c, 0);
-		for (int i = 0; cit != end && (distance < squared_radius); ++cit, ++i) {
-			if ((*cit) < rectangle.min_coord(i))
-				distance += 
-				(rectangle.min_coord(i)-(*cit))*(rectangle.min_coord(i)-(*cit));
-			if ((*cit) > rectangle.max_coord(i))
-				distance += 
-				((*cit)-rectangle.max_coord(i))*((*cit)-rectangle.max_coord(i));
-		}
-		
-		return (distance < squared_radius);
-	}
-
-
-    bool outer_range_contains(const Kd_tree_rectangle<FT,Dimension>& rectangle) const { 
-      // test whether the interior of a sphere
-      // with radius (r+eps) is contained by 'rectangle', i.e.
-      // if the maximal distance of c to the boundary of 'rectangle'
-      // is less than r+eps
-      FT distance=FT(0);
-      FT squared_radius = (r+eps)*(r+eps);	
-      typename SearchTraits::Construct_cartesian_const_iterator_d construct_it=
-          traits.construct_cartesian_const_iterator_d_object();
-      typename SearchTraits::Cartesian_const_iterator_d cit = construct_it(c),
-                                                        end = construct_it(c, 0);
-        for (int i = 0; cit != end && (distance < squared_radius) ; ++cit,++i) {
-		if ((*cit) <= (rectangle.min_coord(i)+rectangle.max_coord(i))/FT(2))
-			distance += 
-			(rectangle.max_coord(i)-(*cit))*(rectangle.max_coord(i)-(*cit));
-		else
-			distance += ((*cit)-rectangle.min_coord(i))*((*cit)-rectangle.min_coord(i));
-		}
-		
-		return (distance < squared_radius);
-	}
-  }; // class Fuzzy_sphere_impl
+namespace internal{
 
+template <class SearchTraits,class Point_d>
+class Fuzzy_sphere_impl
+{
+public:
+  typedef typename SearchTraits::FT FT;
+  typedef typename SearchTraits::Dimension Dimension;
+
+private:
+  SearchTraits traits;
+
+  Point_d c;
+  FT sq_radius;
+  FT sq_inner_radius;
+  FT sq_outer_radius;
+
+public:
+  // default constructor
+  Fuzzy_sphere_impl(const SearchTraits& traits_=SearchTraits()):traits(traits_) {}
+
+  // constructor
+  Fuzzy_sphere_impl(const Point_d& center, FT r, FT eps=FT(0),
+                    const SearchTraits& traits_ = SearchTraits())
+    : traits(traits_), c(center), sq_radius(r*r)
+  {
+    CGAL_precondition(r >= 0);
+    CGAL_precondition(eps >= 0);
+
+    sq_inner_radius = (eps > r) ? 0 : (r - eps) * (r - eps);
```

Comment:
IIUC, this means that you force the inner approximation never to be empty. If we made it empty (for instance using -1 instead of 0 as sq_inner_radius), we could return false more often from inner_range_intersects.

Reply:
But then, if you take r=eps=0 (that is, a non-fuzzy point), the point itself will not intersect its inner range and you will not find it.

Reply:
`0>0` is false, so what I suggested shouldn't change anything for that case. Did I misunderstand your reply?